### PR TITLE
[Go] Heartbeat introduced

### PR DESCRIPTION
**Pull request by ctwarog-ross**
## 📖 Description

Closes Issue #1318

Heartbeat pattern matches cpp
Can be interacted with through server api with Start/StopHeartbeat Can also be interacted with manually as detailed in Heartbeat.go comments

Every specified interval, a given param update will be sent to the client

---

## ✅ Definition of Done (DoD)
- [x] Feature is implemented and works as expected
- [x] Edge cases are handled
- [x] Tests are added or updated (if applicable)
- [x] Documentation is updated (if needed)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes

---

## 🔗 Related Issues / Links

---
## 📝 Additional Context

---

## 🚧 Risks / Open Questions

---
## ✨ Updates
<!-- (Optional) List summarizing changes made when editing the pull request>

### CODE REVIEW

**Review by NR-RV — DISMISSED** [review]

**Comment by ctwarog-ross** [comment]
bugbot run

**Comment by ctwarog-ross** [comment]
bugbot run

File: sdks/go/pkg/catena/heartbeat.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.handlers = append(h.handlers, handler)
+	return h
+}
+
+// Start begins emitting tick events at the specified interval.
+// Returns true if the heartbeat was started, false if it was already running
+// or the interval is invalid (zero or negative).
+func (h *Heartbeat) Start(interval time.Duration) bool {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/heartbeat.go]: I think more go idomatic is to return err if there's an error. and the do nothing if already running, does not need a special return? So that can just return nil, as can the final return and the the if bad internal can do the fmt.Errorf() or w/e that's called

File: sdks/go/pkg/catena/heartbeat.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	h.mu.Lock()
+	handlers := make([]HeartbeatHandler, len(h.handlers))
+	copy(handlers, h.handlers)
+	h.mu.Unlock()
+
+	for _, handler := range handlers {
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					slog.Error("panic in heartbeat handler", "error", r)
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/heartbeat.go]: Should use our logger?

File: sdks/go/pkg/internal/base_server.go [hunk 3]
```diff
@@ -209,3 +211,73 @@ func (bs *BaseServer) BroadcastUpdate(slot uint16, oid string, value any) {
 	}
 	bs.NotifyUpdate(update)
 }
+
```
Comment (abrown-RV) [inline sdks/go/pkg/internal/base_server.go:257]: coverage on the new code please

File: sdks/go/pkg/catena/heartbeat.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	return &Heartbeat{
+		handlers: make([]HeartbeatHandler, 0),
+	}
+}
+
+// OnTick registers a handler to be called on each heartbeat tick.
+// Multiple handlers can be registered and will be called in order.
+// Handlers should not block for long periods.
+// Returns the Heartbeat for method chaining.
+func (h *Heartbeat) OnTick(handler HeartbeatHandler) *Heartbeat {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/heartbeat.go]: should there be a way to remove handlers? I guess you just remake the heartbeat without the handler you don't want anymore....

File: sdks/go/examples/getDevice/getdevice.go [hunk 5]
```diff
@@ -54,6 +56,19 @@ var Devices = map[uint16]map[string]any{
 		"access_scopes":     []string{"st2138:mon", "st2138:op", "st2138:cfg", "st2138:adm"},
 		"default_scope":     "st2138:op",
 		"params": map[string]any{
+			"version": map[string]any{
```
Comment (abrown-RV) [inline sdks/go/examples/getDevice/getdevice.go]: missing the nested `product` struct. Dunno how much we care... the heartbeat doesn't care... but maybe attacted clients will? Unsure here... This is the joy of the detached nature of the handlers architecture...

File: sdks/go/examples/getDevice/getdevice.go [hunk 6]
```diff
@@ -436,15 +452,21 @@ func RegisterHandlers(srv catena.CatenaServer) {
...
+				return catena.ReplyError[catena.CatenaDevice](catena.INTERNAL, "failed to convert device")
 			}
 
 			return catena.Reply(device)
 		})
 	}
+
+	// Start heartbeat — broadcasts the version param on slot 0 every 5 seconds.
+	// This mirrors the C++ SDK pattern: dm.setHeartbeatParam("/product/version"); dm.startHeartbeat();
+	srv.StartHeartbeat(0, "/product/version", func() any {
```
Comment (abrown-RV) [inline sdks/go/examples/getDevice/getdevice.go]: Should tick for each slot. Heartbeat is not just a keeping the connection alive concern. It is also reassuring attached clients that each slot is still ok.

**Review by abrown-RV — CHANGES_REQUESTED** [review]

File: sdks/go/examples/getDevice/getdevice.go (hunk 5 above)
Comment (ctwarog-ross) [reply sdks/go/examples/getDevice/getdevice.go]: Ill make the product struct, to show intended convention

**Comment by ctwarog-ross** [comment]
bugbot run

File: sdks/go/examples/getDevice/getdevice.go [hunk 7]
```diff
@@ -54,6 +56,29 @@ var Devices = map[uint16]map[string]any{
...
+					"version": map[string]any{
+						"name": map[string]any{
+							"display_strings": map[string]string{
+								"en": "Version",
+							},
+						},
+						"type":      catena.ParamTypeString,
+						"read_only": true,
+						"widget":    "TEXT",
+						"value": map[string]any{
```
Comment (abrown-RV) [inline sdks/go/examples/getDevice/getdevice.go]: values in sub params of structs belong at the top level, so the `product` level should have:

```go
"value": map[string]any {
    "struct_value": map[string]any {
        "fields": map[string]any {
            "version": map[string]any {
                "string_value": "1.0.0"
            }
        }
    }
}
```
verbose, I know XD

**Comment by ctwarog-ross** [comment]
bugbot run

**Review by abrown-RV — DISMISSED** [review]

**Review by abrown-RV — APPROVED** [review]
